crates/runtime/src/auth/mod.rs

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use crate::{
+    secrets::{ParamStr, Secrets},
+    App,
+};
+use runtime_auth::{api_key::ApiKeyAuth, FlightBasicAuth, GrpcAuth, HttpAuth};
+use secrecy::ExposeSecret;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+pub struct EndpointAuth {
+    pub http_auth: Option<Arc<dyn HttpAuth + Send + Sync>>,
+    pub flight_basic_auth: Option<Arc<dyn FlightBasicAuth + Send + Sync>>,
+    pub grpc_auth: Option<Arc<dyn GrpcAuth + Send + Sync>>,
+}
+
+impl EndpointAuth {
+    #[must_use]
+    pub async fn new(secrets: Arc<RwLock<Secrets>>, app: &App) -> Self {
+        Self {
+            http_auth: http_auth(Arc::clone(&secrets), app).await,
+            flight_basic_auth: flight_basic_auth(Arc::clone(&secrets), app),
+            grpc_auth: grpc_auth(Arc::clone(&secrets), app),
+        }
+    }
+
+    #[must_use]
+    pub fn no_auth() -> Self {
+        Self {
+            http_auth: None,
+            flight_basic_auth: None,
+            grpc_auth: None,
+        }
+    }
+
+    #[must_use]
+    pub fn with_http_auth(mut self, auth: Arc<dyn HttpAuth + Send + Sync>) -> Self {
+        self.http_auth = Some(auth);
+        self
+    }
+}
+
+impl Default for EndpointAuth {
+    fn default() -> Self {
+        Self::no_auth()
+    }
+}
+
+/// Gets the HTTP auth provider configured for the app, if any
+#[must_use]
+async fn http_auth(
+    secrets: Arc<RwLock<Secrets>>,
+    app: &App,
+) -> Option<Arc<dyn HttpAuth + Send + Sync>> {
+    let api_key_auth = app
+        .runtime
+        .auth
+        .as_ref()
+        .and_then(|auth| auth.api_key.as_ref())?;
+
+    let secrets = secrets.read().await;
+    let mut keys = Vec::with_capacity(api_key_auth.keys.len());
```

Comment:
given this auth check runs on every request, is it possible to make it a e.g. `HashSet` instead of `Vec`?

Reply:
For small numbers of elements, iterating over an array is much faster than computing a hash